x-pack/plugins/cloud_security_posture/public/pages/rules/rules_container.tsx

```diff
@@ -0,0 +1,187 @@
+  changedRules: Map<string, RuleSavedObject>,
+  query: RulesQuery
+): RulesPageData => {
+  const rules = data?.savedObjects || [];
+  const page = getPage(rules, query);
+  return {
+    loading: status === 'loading',
+    error: error ? extractErrorMessage(error) : undefined,
+    all_rules: rules,
+    rules_map: new Map(rules.map((rule) => [rule.id, rule])),
+    rules_page: page.map((rule) => changedRules.get(rule.attributes.id) || rule),
+    total: data?.total || 0,
+  };
+};
+
+const getPage = (data: readonly RuleSavedObject[], { page, perPage }: RulesQuery) =>
+  data.slice(page * perPage, (page + 1) * perPage);
+
+const MAX_ITEMS_PER_PAGE = 10000;
+
+export const RulesContainer = () => {
+  const tableRef = useRef<EuiBasicTable>(null);
+  const [changedRules, setChangedRules] = useState<Map<string, RuleSavedObject>>(new Map());
+  const [isAllSelected, setIsAllSelected] = useState<boolean>(false);
+  const [visibleSelectedRulesIds, setVisibleSelectedRulesIds] = useState<string[]>([]);
+  const [rulesQuery, setRulesQuery] = useState<RulesQuery>({ page: 0, perPage: 5, search: '' });
+
+  const { data, status, error, refetch } = useFindCspRules({
+    search: getSimpleQueryString(rulesQuery.search),
+    page: 1,
+    perPage: MAX_ITEMS_PER_PAGE,
+  });
+
+  const { mutate: bulkUpdate, isLoading: isUpdating } = useBulkUpdateCspRules();
+
+  const rulesPageData = useMemo(
+    () => getRulesPageData({ data, error, status }, changedRules, rulesQuery),
+    [data, error, status, changedRules, rulesQuery]
+  );
+
+  const hasChanges = !!changedRules.size;
+
+  const selectAll = () => {
+    if (!tableRef.current) return;
+    tableRef.current.setSelection(rulesPageData.rules_page);
+    setIsAllSelected(true);
+  };
+
+  const toggleRules = (rules: RuleSavedObject[], enabled: boolean) =>
+    setChangedRules(
+      getChangedRules(
+        rulesPageData.rules_map,
+        changedRules,
+        rules.map((rule) => ({
+          ...rule,
+          attributes: { ...rule.attributes, enabled },
+        }))
+      )
+    );
+
+  const bulkToggleRules = (enabled: boolean) =>
+    toggleRules(
+      isAllSelected
+        ? rulesPageData.all_rules
+        : visibleSelectedRulesIds.map((ruleId) => rulesPageData.rules_map.get(ruleId)!),
+      enabled
+    );
+
+  const toggleRule = (rule: RuleSavedObject) => toggleRules([rule], !rule.attributes.enabled);
+
+  const bulkUpdateRules = () => bulkUpdate([...changedRules].map(([, rule]) => rule.attributes));
+
+  const discardChanges = useCallback(() => setChangedRules(new Map()), []);
+
+  const clearSelection = useCallback(() => {
+    if (!tableRef.current) return;
+    tableRef.current.setSelection([]);
+    setIsAllSelected(false);
+  }, []);
+
+  useEffect(discardChanges, [data, discardChanges]);
+  useEffect(clearSelection, [rulesQuery, clearSelection]);
+
+  return (
+    <div data-test-subj={TEST_SUBJECTS.CSP_RULES_CONTAINER}>
```

Comment:
You don't need a DOM node here, you can pass the test subject to the panel and use a fragment

Reply:
that div is there to separate the panel and the bottom bar

Reply:
What do you mean by "separate"?

Reply:
as in the div is the container of the panel and the bottom bar (the one with the "save" button)

if i place the bottom bar in the panel it changes its height (jumping around)

i guess we can solve this in other ways too. but IMO this is simple enough